src/main/java/io/nats/client/impl/NatsConnection.java

```diff
@@ -610,39 +613,45 @@ void handleCommunicationIssue(Exception io) {
     // Close socket is called when another connect attempt is possible
     // Close is called when the connection should shutdown, period
     void closeSocket(boolean tryReconnectIfConnected) throws InterruptedException {
-        boolean wasConnected;
+        // Ensure we close the socket exclusively within one thread.
+        closeSocketLock.lock();
 
-        statusLock.lock();
         try {
-            if (isDisconnectingOrClosed()) {
-                waitForDisconnectOrClose(this.options.getConnectionTimeout());
-                return;
+            boolean wasConnected;
+
+            statusLock.lock();

```

Comment:
Do we need both the closeSocketLock and the statusLock, or can we just rely on the closeSocketLock?

Reply:
It's needed yes.
Both removing the `statusLock`, solely relying on `closeSocketLock` fails the tests. As well as wrapping the `statusLock` around it fully, instead of using the `closeSocketLock`.